feat(env-check): add environment check for vscode

cianmSAP opened the pull request:
#694

Add env check for vscode

tobiasqueck commented on packages/environment-check/package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@sap-ux/environment-check",
-    "version": "0.1.4",
+    "version": "0.1.5",
```
Please leave the version number as-is. Once you are done, you add a changeset and the version number will be updated automatically.

cianmSAP replied:
Updated, thanks @tobiasqueck

Klaus-Keller commented on packages/environment-check/src/cli/index.ts:
```diff
@@ -150,7 +151,7 @@ export async function cli(): Promise<void> {
         const options = getOptions(cliArgs) || {};
         options.credentialCallback = credentialCallback;
-        const result = await checkEnvironment(options);
+        const result = isAppStudio() ? await checkBASEnvironment(options) : await checkVSCodeEnvironment();
```
as mentioned above, I'd keep `checkEnvironment()` as is and move the check done here inside `checkEnvironment`.

cianmSAP replied:
Done

Klaus-Keller commented on packages/environment-check/src/checks/index.ts:
```diff
@@ -1,2 +1,2 @@
-export { checkEnvironment, getEnvironment } from './environment';
+export { checkBASEnvironment, getBASEnvironment, checkVSCodeEnvironment, getVSCodeEnvironment } from './environment';
```
same as above, please keep the public interface 
```
export { checkEnvironment, getEnvironment } from './environment';
```
stable

Klaus-Keller commented on packages/environment-check/src/checks/install.ts:
```diff
@@ -0,0 +1,64 @@
+ * @param module - module id
+ * @returns version
+ */
+export const getCFCliToolVersion = async (module: string): Promise<string> => {
```
The name suggests that this function returns the version of the "Clound Foundry CLI (cf)", but you still need to pass parts of the CLI command into it. I'd say, either this function returns the version of cf and doesn't require the consumer to know about the command name (`cf`), or we rename this function to indicate, that you can pass any command line command and it will return the version (as long as argument -v returns the version. I'm in favor of the first and made the suggestions accordingly.

```suggestion
export const getCFCliToolVersion = async (): Promise<string> => {
```

cianmSAP replied:
Updated as suggested

Klaus-Keller commented on packages/environment-check/src/checks/install.ts:
```diff
@@ -0,0 +1,64 @@
+export const getCFCliToolVersion = async (module: string): Promise<string> => {
+    let cfVersion: string;
+    try {
+        const ouput = await spawnCommand(module, ['-v']);
```
```suggestion
        const ouput = await spawnCommand(NpmModules.CloudCliTools, ['-v']);
```
Requires also to add 
```
import { NpmModules } from '../types';
```
to the import section. 

Also there is a type: ouput instead of output

cianmSAP replied:
Updated as suggested

Klaus-Keller commented on packages/environment-check/src/checks/install.ts:
```diff
@@ -0,0 +1,64 @@
+import { t } from '../i18n';
+import { join } from 'path';
+import { spawnCommand, npmCommand } from '../command';
+
```
Can we use format
[export] function <FUNCTION_NAME>() {
<FUNCTION_BODY>
}
for top level functions to be consistent with the rest of this module? 
Usually we use arrow function expressions for anonymous functions or functions that do not require a body.

Additionally, the name `install.ts` is a bit misleading.

cianmSAP replied:
Updated

Klaus-Keller commented on packages/environment-check/src/checks/install.ts:
```diff
@@ -0,0 +1,64 @@
+ * @returns version
+ */
+export const getFioriGenVersion = async (module: string): Promise<string> => {
+    let fioriGenVersion: string;
```
similar comment like for `getCFCliToolVersion()`, I'd either remove the need to add the `module` name or make this function more generic so it can return the version of any globally installed module and rename it.

cianmSAP replied:
Updated

Klaus-Keller commented on packages/environment-check/src/output/markdown.ts:
```diff
@@ -3,12 +3,14 @@ import { Severity, UrlServiceType } from '../types';
+    VSCodeEnvironment
 } from '../types';
 import { t } from '../i18n';
+import { isAppStudio } from '@sap-ux/btp-utils';
```
Ideally, the markdown writer doesn't have any knowledge about the environment. We should try to find a result structure that can be converted to markdown without knowing which environment we currently running in.

cianmSAP replied:
See comment

Klaus-Keller commented on packages/environment-check/src/output/zip.ts:
```diff
@@ -25,8 +25,13 @@ function byteNumberToSizeString(byteNumber: number): string {
+export function storeResultsZip(
+    results: EnvironmentCheckResult,
+    targetFile = 'envcheck-results.zip',
+    markdownOnly = false
```
In which case do we need `markdownOnly`? I think if we store check results as zip it is always better to include json and markdown. @theStiggySpd was there a specific request?

cianmSAP replied:
I did raise this in grooming and it was requested to provide markdown only.

cianmSAP commented on packages/environment-check/src/types.ts:
```diff
@@ -14,11 +14,41 @@ export enum OutputMode {
+interface Environment {
     platform: NodeJS.Platform;
-    versions: NodeJS.ProcessVersions;
+}
```
It wasn't removed as such but moved into BASEnviornment and there was no request to include the additional `NodeJS.ProcessVersions` (`v8`, `uv`, `zlib` etc ) in the VSCode results.

cianmSAP commented:
With the ask to include `Gather environment information` for BAS also as per https://github.wdf.sap.corp/ux-engineering/tools-suite/issues/21564#issue-5803089 I've exposed the `getToolsExtensions` function and also reverted the checkEnvironment and getEnvironment functions as requested.

The only slight troube with keeping the markdown file unaware of the dev environment is/was the title. I've added the markdownTitle property to the `EnvironmentCheckResult` interface and exposed the `getMarkdownTitle` function to retreieve it. It may be over engineering a simple part so please suggest an appropriate or alternative way. 
Thanks @Klaus-Keller

cianmSAP commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -8,31 +8,33 @@ import type {
+ * @returns environment, including platform, versions ...
  */
-export async function getEnvironment(): Promise<{ environment: Environment; messages: ResultMessage[] }> {
+export async function getBASEnvironment(): Promise<{ environment: BASEnvironment; messages: ResultMessage[] }> {
```
Done

Klaus-Keller commented on packages/environment-check/src/types.ts:
```diff
@@ -22,6 +36,20 @@ export interface Environment {
     basDevSpace?: string;
 }
 
+export interface ToolsExtensions {
```
Interface `ToolsExtensions` looks good, but I would have an optional property in `Environment` with this interface. Logically extension and module versions belong to the environment. Right now we read and fill this information when running in VSCode only, but at some point we might collect the same information in BAS.

cianmSAP replied:
Updated

Klaus-Keller commented on packages/environment-check/src/output/markdown.ts:
```diff
@@ -259,11 +296,15 @@ function writeMessages(writer: MarkdownWriter, messages: ResultMessage[] = []):
-    writeDestinations(writer, results.destinations);
-    writeMessages(writer, results.messages);
+    writer.addH1(results.markdownTitle);
+    if (results.toolsExtensions) {
```
Can we avoid having an `if` here and output whatever is there in the `EnvironmentCheckResult`. We might fill certain properties only in certain cases, but the writer should just traverse over the results and output markdown, not aware of other logic. This keeps the writer simple. I assume at some point we might also check certain module versions in BAS and destinations in VSCode.

Klaus-Keller commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -137,31 +204,44 @@ async function getDestinationsResults(
+        logger.push(...destResults.messages);
+    }
+
+    const toolsExtensionResults = await getToolsExtensions();
```
I've tested in BAS and when I enter this function I get an error:

```
Error checking environment: Error: spawn code ENOENT
```

Because command `code` doesn't exist in BAS. This indicates also that we need a better error handling, if getting one installed extension/module fails, it should continue with the others and not completely exit.

cianmSAP replied:
Fixed

Klaus-Keller commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -44,6 +48,69 @@ export async function getEnvironment(): Promise<{ environment: Environment; mess
+}> {
+    const logger = getLogger();
+
+    const extensions = await getInstalledExtensions();
```
this step could fail, but should not cause the whole check to fail. Perhaps wrap it in `try/catch` or add try catch inside `getInstalledExtensions`

Klaus-Keller reviewed (changes requested):
Hi @cianmSAP, 
See my comments with further review. I did not review the whole code yet, will continue tomorrow.

cianmSAP commented on packages/environment-check/src/output/markdown.ts:
```diff
@@ -254,15 +294,18 @@ function writeMessages(writer: MarkdownWriter, messages: ResultMessage[] = []):
  * @returns - markdown report
  */
-export function convertResultsToMarkdown(results: EnvironmentCheckResult): string {
+export function convertResultsToMarkdown(results: EnvironmentCheckResult, isDestinationsChecked = true): string {
```
I've added this as I don't want to implement a check for the existence of destinations in the results, as I've seen in support where customers have no destinations in the results due to configuration issues. 
Therefore I think it might be better to pass the param from app-mod when there has actually been no checked performed.

Klaus-Keller replied:
Good consideration, we need a way to differentiate between:
- 0 destinations due to configuration issues
- 0 destinations because we just do an environment check

We could even extend this concept a bit, instead of manually passing this into `convertResultsToMarkdown()` function, could we add a property on the result state that indicates which check were requested? 

E.g. extend type  `EnvironmentCheckResult`
https://github.com/SAP/open-ux-tools/blob/e026fc2f82327df8068c7c4e29f874bf05528eab/packages/environment-check/src/types.ts#L65-L70
with

```typescript
    requestedChecks: Set<Check>
}

type Check = 'environment' | 'destination' | 'destinationResults';

```
We also allow to skip `destinationResults` and the same concept could be applied there. 
That way it would be clear from the results which check was executed and the markdown writer could skip results for not executed checks.

What do you think?

cianmSAP replied:
Implemented this, I think it will keep the markdown clean when the check has not been performed. Thanks @Klaus-Keller

Klaus-Keller commented:
> The only slight troube with keeping the markdown file unaware of the dev environment is/was the title. I've added the markdownTitle property to the `EnvironmentCheckResult` interface and exposed the `getMarkdownTitle` function to retrieve it. It may be over engineering a simple part so please suggest an appropriate or alternative way. Thanks @Klaus-Keller

@theStiggySpd, why do we need different title for VSCode or BAS results? This information is already included in "Environment", see:

![image](https://user-images.githubusercontent.com/66327622/193694002-1c9a9d9b-cbec-4223-a64f-dd008dbd4b43.png)

Personally, I think we should try to have the checks as development environment agnostic as possible, what do you think?

Klaus-Keller commented:
@cianmSAP, another information, there is in new version of `@sap/bas-sdk` available: `2.1.1`. The new version includes functionality to get the devspace information. 
We can delete our custom function `getSbasDevspace()`:

https://github.com/SAP/open-ux-tools/blob/e026fc2f82327df8068c7c4e29f874bf05528eab/packages/environment-check/src/checks/environment.ts#L169-L190


and retrieve the information directly from `bas-sdk`, like:

```
import { core as basCoreApi, devspace } from '@sap/bas-sdk';

..
      if (isAppStudio()) {
          environment.basDevSpace = (await devspace.getDevspaceInfo()).packDisplayName;

```

cianmSAP commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -27,23 +29,113 @@ export async function getEnvironment(): Promise<{ environment: Environment; mess
+ * @param extensions to be checked
+ * @returns extension with version numbers
+ */
+function __getExtVersions(extensions: { [id: string]: { version: string } }): { [id: string]: string } {
```
Updated

cianmSAP commented on packages/environment-check/src/checks/getInstalled.ts:
```diff
@@ -0,0 +1,154 @@
+ *
+ * @returns list of extension ids and versions
+ */
+async function _getExtensionsBAS(): Promise<{ [id: string]: { version: string } }> {
```
Updated

cianmSAP commented on packages/environment-check/src/checks/getInstalled.ts:
```diff
@@ -0,0 +1,154 @@
+ *
+ * @returns version
+ */
+async function _getFioriGenBAS(): Promise<string> {
```
Updated

cianmSAP commented on packages/environment-check/src/checks/getInstalled.ts:
```diff
@@ -0,0 +1,154 @@
+ *
+ * @returns version
+ */
+async function _getFioriGenGlobalRoot(): Promise<string> {
```
Updated

Klaus-Keller commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -137,26 +219,32 @@ async function getDestinationsResults(
+    let destinations: Destination[];
+    let destinationResults: { [dest: string]: DestinationResults };
 
     const { environment, messages } = await getEnvironment();
```
we now need to set the ``requestedChecks` otherwise they are not part of the markdown if you run it from command line:

```suggestion
    const requestedChecks: Set<Check> = new Set();
    
    requestedChecks.add(Check.Environment);
    const { environment, messages } = await getEnvironment();
```

cianmSAP replied:
Thanks Klaus, slipped my mind that we could generate the markdown from the CLI. Updated now

Klaus-Keller commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -137,26 +219,32 @@ async function getDestinationsResults(
-        options?.credentialCallback
-    );
-    logger.push(...destDetailsMessages);
+        const basDestResults = await checkBASDestinations();
```
similar here, when we check destinations we need to add it to requested checks

```suggestion
        requestedChecks.add(Check.Destinations);
        const basDestResults = await checkBASDestinations();
```

cianmSAP replied:
Updated

Klaus-Keller commented on packages/environment-check/src/checks/environment.ts:
```diff
@@ -137,26 +219,32 @@ async function getDestinationsResults(
+        destinations = basDestResults.destinations;
+        logger.push(...basDestResults.messages);
+
+        const destResults = await getDestinationsResults(
```
and here, proposal:

```suggestion
        if (deepDiveDestinations.size > 0) {
            requestedChecks.add(Check.DestResults);
        }
        const destResults = await getDestinationsResults(
```

cianmSAP replied:
Updated

Klaus-Keller reviewed (changes requested):
@cianmSAP, looks pretty good. I found one small issue when running the check from command line now, see comments, should be easy fix.

Klaus-Keller commented on packages/environment-check/src/output/zip.ts:
```diff
@@ -20,6 +21,32 @@ function byteNumberToSizeString(byteNumber: number): string {
+ * @param results - environment check results
+ * @returns json string
+ */
+function convertResultsToJson(results: EnvironmentCheckResult): string {
```
not sure whether we need to massage the result json. I think I would keep it raw, it contains the information of `requestedChecks` so when checking the JSON it is should be clear what checks have been executed

cianmSAP replied:
Ok. I've reverted

Klaus-Keller commented:
Hi @cianmSAP, sorry, I forgot that a set is not serialized when running `JSON.stringify()`. This is why the `requestedChecks` where not shown in the resulting json. 
There are two ways to overcome this:
1. Write an own serializer for `Set`, e.g. by passing a function as second argument to `JSON.stringify`
2. Use array instead of Set

I think it easier to switch to array, so I pushed a commit: https://github.com/SAP/open-ux-tools/pull/695/commits/08a15c5e9ca93e301d446c174d9db0c470447d91

Klaus-Keller reviewed (approved):
- Checked code and it looks good to me now
- Test coverage is excellent
- Build and tested via CLI in VScode and SBAS, no issues found

**Approved**. Thanks a lot @cianmSAP

Klaus-Keller commented:
@tobiasqueck, could you please also check again.

tobiasqueck reviewed (approved):
Not required to be changed for this PR, however, you should think about an object oriented approach with a base class and extensions for VSCode and BAS. That would remove requiring `isAppStudio` checks in e.g. `getFioriGenVersion` or `getInstalledExtensions`.

Also the interfaces e.g. `ToolsExtensions` would be easier to read if the properties had a short explanation each.

All in all, the code is easy to read, and the tests are covering the functionality, therefore, please go ahead and merge. Nevertheless, please consider my above comments in a follow up PR.

cianmSAP commented:
> Not required to be changed for this PR, however, you should think about an object oriented approach with a base class and extensions for VSCode and BAS. That would remove requiring `isAppStudio` checks in e.g. `getFioriGenVersion` or `getInstalledExtensions`.
> 
> Also the interfaces e.g. `ToolsExtensions` would be easier to read if the properties had a short explanation each.
> 
> All in all, the code is easy to read, and the tests are covering the functionality, therefore, please go ahead and merge. Nevertheless, please consider my above comments in a follow up PR.

Noted, thanks @tobiasqueck

Klaus-Keller commented:
> Not required to be changed for this PR, however, you should think about an object oriented approach with a base class and extensions for VSCode and BAS. That would remove requiring `isAppStudio` checks in e.g. `getFioriGenVersion` or `getInstalledExtensions`.

@tobiasqueck, let's have a chat about this before adding classes into this functional component.

tobiasqueck commented:
> let's have a chat about this before adding classes into this functional component.

Sure, we can discuss. It see it as a simple "is there a state or not" question, and by looking at the code, I see more than a handful `isAppStudio` calls leading to `if-else` conditions that strongly indicate that using polymorphism here would reduce the cognitive complexity of the code. It would also cleanly separate environment specific code into one place instead of mixing it everywhere.

Klaus-Keller commented:
> Sure, we can discuss. It see it as a simple "is there a state or not" question, and by looking at the code, I see more than a handful `isAppStudio` calls leading to `if-else` conditions that strongly indicate that using polymorphism here would reduce the cognitive complexity of the code. It would also cleanly separate environment specific code into one place instead of mixing it everywhere.

Yes, fully agree to improve the separation `isAppStudio` or not, ideally the differences should be minimal and clearly visible in the structure of the code. We can certainly improve this.
